### Importing Libraries

In [1]:

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score

### Loading Data

In [2]:
X=pd.read_csv("/kaggle/input/hit-prediction-processed-data/Hit Prediction/X.csv")
X_train=pd.read_csv("/kaggle/input/hit-prediction-processed-data/Hit Prediction/X_train.csv")
X_test=pd.read_csv("/kaggle/input/hit-prediction-processed-data/Hit Prediction/X_test.csv")
X_selected=pd.read_csv("/kaggle/input/hit-prediction-processed-data/Hit Prediction/X_selected.csv")
X_train_selected=pd.read_csv("/kaggle/input/hit-prediction-processed-data/Hit Prediction/X_train_selected.csv")
X_test_selected=pd.read_csv("/kaggle/input/hit-prediction-processed-data/Hit Prediction/X_test_selected.csv")
y=np.load("/kaggle/input/hit-prediction-processed-data/Hit Prediction/y.npy")
y_train=np.load("/kaggle/input/hit-prediction-processed-data/Hit Prediction/y_train.npy")
y_test=np.load("/kaggle/input/hit-prediction-processed-data/Hit Prediction/y_test.npy")
y_selected=np.load("/kaggle/input/hit-prediction-processed-data/Hit Prediction/y_selected.npy")

### Fitting Basic Model with default Parameters

#### Building Base Model

In [3]:
clf = KNeighborsClassifier(n_neighbors=5)
clf.fit(X_train_selected, y_train)
clf.score(X_test_selected, y_test)

0.7424345626155493

### Cross Validation Score on Basic Model

In [4]:
scores = cross_val_score(clf, X_selected, y_selected, cv=5)
scores

array([0.74288494, 0.73737988, 0.74236711, 0.74638122, 0.73701496])

In [5]:
print("%0.5f accuracy with a standard deviation of %0.5f" % (scores.mean(), scores.std()))

0.74121 accuracy with a standard deviation of 0.00355


### Finding Best Parameters using Grid Search

#### Hyper Parameter List

In [6]:
parameters = {
'n_neighbors':[3,5,7,9,11,13,15,17,19],
'weights':[ "uniform", "distance"],
 'algorithm' : ["auto", "ball_tree", "kd_tree", "brute"],   
'p':[1,2]
}

#### Base Model

In [7]:
clf=KNeighborsClassifier()

#### Grid Search

In [ ]:
gcv = GridSearchCV(clf, parameters, cv=5, verbose=False, n_jobs=-1 )
gcv.fit(X_train_selected,y_train)

In [ ]:
gcv.best_params_

#### Grid Search Cross Validation Mean Score

In [ ]:
gcv.best_score_

In [ ]:
gcv.best_estimator_

### Test Accuracy for Model with best hyperparameters

In [ ]:
clf=gcv.best_estimator_
clf.fit(X_train_selected,y_train)
clf.score(X_test_selected,y_test)

In [ ]:
pred=clf.predict(X_test_selected)
print(confusion_matrix(y_test,pred))

In [ ]:
print(classification_report(y_test,pred))

### Cross Validation Score on best model 

In [ ]:
scores = cross_val_score(clf, X_selected, y_selected, cv=5)
scores

In [ ]:
print("%0.5f accuracy with a standard deviation of %0.5f" % (scores.mean(), scores.std()))

#### Inference -
1. The cross validation score for the base model with default parameters is **0.74121**. 

2. The cross validation score for the best model with hyper parameters {'algorithm': 'ball_tree', 'n_neighbors': 19, 'p': 1, 'weights': 'distance'} is **0.76473**.